### PR TITLE
Delegate

### PR DESCRIPTION
**Pull request by lclc98**
Delegate is a way to have calls run on a certain side

### CODE REVIEW

File: src/main/java/xyz/openmodloader/modloader/ManifestModContainer.java [hunk 1]
```diff
@@ -117,7 +125,8 @@ public ResourceLocation getLogoTexture() {
             try {
                 InputStream in = null;
                 if (logoBytes == null) {
-                    in = new URL(getModFile().toURI().toURL().toString() + '/' + logo).openStream();;
+                    in = new URL(getModFile().toURI().toURL().toString() + '/' + logo).openStream();
+                    ;
```
Comment (phase) [inline src/main/java/xyz/openmodloader/modloader/ManifestModContainer.java]: What is this?

File: src/main/java/xyz/openmodloader/modloader/ModLoader.java [hunk 2]
```diff
@@ -62,7 +60,7 @@
     /**
      * Attempts to load all mods from the mods directory and the classpath. While this is public,
      * it is intended for internal use only!
-     * 
+     * <p>
```
Comment (phase) [inline src/main/java/xyz/openmodloader/modloader/ModLoader.java]: Add a `</p>` at the end.

File: src/main/java/xyz/openmodloader/modloader/ModLoader.java [hunk 3]
```diff
@@ -106,6 +105,23 @@ public static void loadMods() {
...
+                            try {
+                                field.setAccessible(true);
+                                if (Modifier.isFinal(field.getModifiers())) {
+                                    Field modifiersField = Field.class.getDeclaredField("modifiers");
+                                    modifiersField.setAccessible(true);
+                                    modifiersField.setInt(field, field.getModifiers() & ~Modifier.FINAL);
+                                }
+                                field.set(instance, delegate);
+                            } catch (NoSuchFieldException | SecurityException | IllegalArgumentException | IllegalAccessException e) {
+                                OpenModLoader.getLogger().error("Could not set @Instance field", e);
```
Comment (phase) [inline src/main/java/xyz/openmodloader/modloader/ModLoader.java]: Should this say `@Delegate`?

File: src/test/resources/META-INF/MANIFEST.MF [hunk 4]
```diff
@@ -7,3 +7,5 @@ Author: OML
 URL: http://example.com/
 Update-URL: https://gist.githubusercontent.com/anonymous/bf49748c02fc854be2cc312110b91ebc/raw/ad770f1c717a4d22eb6e89984fec8cdcffe560ea/2016-06-26_20-53-08.txt
 Logo: logo.png
+Client-Delegate: xyz.openmodloader.test.OMLClientDelegate
+Server-Delegate: xyz.openmodloader.test.OMLServerDelegate
```
Comment (phase) [inline src/test/resources/META-INF/MANIFEST.MF]: Are these manual?

**Comment by phase** [comment]
Are these synonymous with Forge's proxies? (No clue if I'm talking about the right thing.)

**Comment by lclc98** [comment]
Yeah

**Comment by Victorious3** [comment]
I've always found Forge's proxies especially useless since you can achieve the same thing with one if statement and a method call.

**Comment by phase** [comment]
Using lambda instead of separate classes is an idea. Maybe not a _good_ one, but it's better than two extra classes that you have to write down the paths of.

@Victorious3 I think separating client and server code helps somewhat. Rewriting `if (side == Side.CLIENT)` a billion times is error prone. Overriding client methods is cleaner.

**Comment by Victorious3** [comment]
@phase Yeah I agree which is why I'd write exactly *one* if statement and put all my code in that one. It is much cleaner than having a couple dozen of empty methods you have to write for Forge's proxies, plus the need to add two more class names to the manifest which have the notorious tendency to break if you decide to rename your packages.

**Comment by MrIbby** [comment]
I like @phase's idea with lamda. It could look like this: `DelegateBuilder.client(OMLClientDelegate::new).server(OMLServerDelegate::new).build()`

**Comment by phase** [comment]
@MrIbby Check Discord, we're going to do `@Delegate(client = ClientDelegate.class, server = ServerDelegate.class)`.

**Comment by Victorious3** [comment]
@MrIbby How's that better than a simple... if statement? This is over-engineering things.

**Comment by MrIbby** [comment]
@Victorious3 I think some people would prefer this:
`Delegates.newDelegate(OMLClientDelegate::new, OMLServerDelegate::new)`
rather than this:
``` java
if (OpenModLoader.getSide() == Side.CLIENT)
    delegate = new OMLClientDelegate();
else
    delegate = new OMLServerDelegate();
```

**Comment by Victorious3** [comment]
@MrIbby No, I meant, what's the whole point of this vs a simple if statement to check if its on the client side? Its... more complicated and uses reflection.

**Comment by phase** [comment]
Instead of having separate classes, we can use `@Strippable` for things we want only on the client side, and an if statement for things that are on the server and client side.

**Comment by darkhax** [comment]
While I like the idea of the delegate system, @Victorious3 and @phase are right, it is obsolete. I am also not very fond of how the example delegate code is set up, it gives the wrong impression.

**Pull request closed without merging**